hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/ozone/container/common/TestEndPoint.java

```diff
@@ -0,0 +1,326 @@
+          rpcEndpoint.getState());
+    }
+  }
+
+
+  @Test
+  public void testRegisterNoContainerID() throws Exception {
+    InetSocketAddress address = SCMTestUtils.getReuseableAddress();
+    try (EndpointStateMachine rpcEndpoint =
+             registerTaskHelper(address, 1000, true)) {
+      // No Container ID, therefore we tell the datanode that we would like to
+      // shutdown.
+      Assert.assertEquals(EndpointStateMachine.EndPointStates.SHUTDOWN,
+          rpcEndpoint.getState());
+    }
+  }
+
+  @Test
+  public void testRegisterRpcTimeout() throws Exception {
+    final long rpcTimeout = 1000;
+    final long tolerance = 200;
+    scmServerImpl.setRpcResponseDelay(1500);
+    long start = Time.monotonicNow();
+    registerTaskHelper(serverAddress, 1000, false).close();
+    long end = Time.monotonicNow();
+    scmServerImpl.setRpcResponseDelay(0);
+    Assert.assertThat(end - start, new LessOrEqual<>(rpcTimeout + tolerance));
+  }
+
+  @Test
+  public void testHeartbeat() throws Exception {
+    DatanodeID dataNode = SCMTestUtils.getDatanodeID();
+    try (EndpointStateMachine rpcEndPoint =
+             SCMTestUtils.createEndpoint(SCMTestUtils.getConf(),
+                 serverAddress, 1000)) {
+      SCMHeartbeatResponseProto responseProto = rpcEndPoint.getEndPoint()
+          .sendHeartbeat(dataNode);
+      Assert.assertNotNull(responseProto);
+      Assert.assertEquals(1, responseProto.getCommandsCount());
+      Assert.assertNotNull(responseProto.getCommandsList().get(0));
+      Assert.assertEquals(responseProto.getCommandsList().get(0).getCmdType(),
+          Type.nullCmd);
+    }
+  }
+
+
+  private EndpointStateMachine heartbeatTaskHelper(InetSocketAddress scmAddress,
+      int rpcTimeout) throws Exception {
+    EndpointStateMachine rpcEndPoint = SCMTestUtils.createEndpoint(
+        SCMTestUtils.getConf(),
+        scmAddress, rpcTimeout);
+    ContainerNodeIDProto  containerNodeID = ContainerNodeIDProto.newBuilder()
+        .setClusterID(UUID.randomUUID().toString())
+        .setDatanodeID(SCMTestUtils.getDatanodeID().getProtoBufMessage())
+        .build();
+    rpcEndPoint.setState(EndpointStateMachine.EndPointStates.HEARTBEAT);
+    HeartbeatEndpointTask endpointTask =
+        new HeartbeatEndpointTask(rpcEndPoint, SCMTestUtils.getConf());
+    endpointTask.setContainerNodeIDProto(containerNodeID);
+    endpointTask.call();
+    Assert.assertNotNull(endpointTask.getContainerNodeIDProto());
+    return rpcEndPoint;
+  }
+
+  @Test
+  public void testHeartbeatTask() throws Exception {
+    try(EndpointStateMachine rpcEndpoint =
+            heartbeatTaskHelper(serverAddress, 1000)) {
+      Assert.assertEquals(EndpointStateMachine.EndPointStates.HEARTBEAT,
+          rpcEndpoint.getState());
+    }
+  }
+
+  @Test
+  public void testHeartbeatTaskToInvalidNode() throws Exception {
+    InetSocketAddress invalidAddress = SCMTestUtils.getReuseableAddress();
+    try(EndpointStateMachine rpcEndpoint =
```

Comment:
This can be wrapped in a help function to avoid duplication among cases in testHeartbeatTask*